### DISCUSSION TITLE
Metric Value from Database & avoid BlockingOperationNotAllowedException

### DISCUSSION THREAD

**Discussion by pepperbob** [quarkusio/quarkus#23605] [Q&A] [answered]
My aim is to build a metric that essentially reads a number from a database table which then should be recorded in a Gauge.

So far I managed to register the metric via a `BuildStep` and a `MetricsFactoryConsumerBuildItem` which is working quite fine - the metric is registered and invoked.

However, when the metric is invoked, the registered `Supplier<Number>` tries execute a query using the `EntityManager` which produces a `BlockingOperationNotAllowedException`. Looking at the MetricFactory API I can't see way how to register something that helps with async. I wanted to avoid polling the database in the background - esp. as it feels quite inconvenient.

Any pointers/ideas how to accomplish this?

**Comment by jmartisk** [comment, marked as answer]
Are you using smallrye-metrics or micrometer? And what exporter? I believe that the Prometheus exporter for both these metric engines is configured to be on a blocking route, which should avoid your issue... Huh.

**Reply by pepperbob** [reply]
Just played around with this:

In my test, when I use the `quarkus-smallrye-metrics` artifact, exporting metrics via HTTP works without a problem - my metric is there, no warning.

When using the `quarkus-micrometer-registry-prometheus` artifact then I'll see the `BlockingOperationNotAllowedException` Warning in the log and my metric is not listed (HTTP Endpoint generally exports metrics, but mine is missing).

🤔

**Reply by pepperbob** [reply]
Okay, found some explanation: with `quarkus-micrometer-registry-prometheus` querying

`/q/metrics`

gives me my metric, no warning. **Querying this as JSON** (what I do in my test 🙈 as it fits RestAssured better) will give me the `BlockingOperationNotAllowedException` warning without my metric.

Opposing `quarkus-smallrye-metrics` where both is possible.

>  I believe that the Prometheus exporter for both these metric engines is configured to be on a blocking route

Ok, this explains why it works for Micrometer with Prometheus; so the difference is basically the different way that the "json exporter" (is this a thing?) is setup for micrometer and smallrye... unexpected though.

**Reply by jmartisk** [reply]
Ok so you've found a bug in the Micrometer JSON exporter. It's using a handler that is not blocking. Prometheus exporter is using a blocking handler, which is correct.
See https://github.com/quarkusio/quarkus/issues/23661
